### PR TITLE
Accessibility Helper Functions

### PR DESCRIPTION
**Pull request by BrettMayson**
**When merged this pull request will:**
- Bring over `epilepsyFriendlyMode` from ACE
- Add options to disable or add a visual hint to sounds
- Helper functions for mods for `playSound`, `playSound3D`, `say2D`, `say3D`

### CODE REVIEW

File: addons/accessibility/fnc_playSound.sqf [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    _sound - Sound to play <STRING>
+    _categories - Categories the sound is in <ARRAY>
+        - "BREATH"
+        - "BODILY"
+        - "PAIN"
+        - "SCREAM"
+        - "FOOD"
+        - "EQUIPMENT"
+        - "REPETITIVE"
+    _args - Arguments passed to playSound3D <ARRAY>
```
Comment (rautamiekka) [inline addons/accessibility/fnc_playSound.sqf:19]: It says 3D here but later doesn't.

File: addons/accessibility/fnc_playSound3D.sqf [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+if (_source isEqualTo objNull) exitWith {false};
+if (_sound isEqualTo "") exitWith {false};
+
+private _shouldPlay = [_categories] call CBA_accessibility_fnc_shouldPlay;
+private _shouldHint = [_categories] call CBA_accessibility_fnc_shouldHint;
+
+if (_shouldPlay) then {
+    _args = [_sound, _source] + _args;
+    playSound3D _args;
```
Comment (PabstMirror) [inline addons/accessibility/fnc_playSound3D.sqf:57]: has global effects
I think always needs to play

Comment (BrettMayson) [reply addons/accessibility/fnc_playSound3D.sqf:57]: Ah right, it does have an argument for localonly, so just need to use an event and let each client decide if they want to play it

File: addons/accessibility/fnc_hint.sqf [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+params [
+    ["_hint", "", ["", []]],
+    ["_source", objNull, [objNull]]
+];
+
+if (_hint isEqualType []) then {
+    _hint
+} else {
+    [_hint]
```
Comment (commy2) [inline addons/accessibility/fnc_hint.sqf:34]: What's this for? `params` takes any lhs.

File: addons/accessibility/fnc_playSound3D.sqf [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    ["_source", objNull, [objNull]],
+    ["_args", [], [[]]],
+    ["_hint", "", ["", []]]
+];
+
+if (_source isEqualTo objNull) exitWith {false};
+if (_sound isEqualTo "") exitWith {false};
+
+private _shouldPlay = [_categories] call CBA_accessibility_fnc_shouldPlay;
+private _shouldHint = [_categories] call CBA_accessibility_fnc_shouldHint;
```
Comment (commy2) [inline addons/accessibility/fnc_playSound3D.sqf:53]: "should"? This is neither an expectation, suggestion or an obligation. You _should_ use descriptive mode:

CBA_accessibility_fnc_soundEnabled, CBA_accessibility_fnc_hintEnabled.

File: addons/accessibility/fnc_vanillaSoundPlayed.sqf [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+Author:
+    Brett Mayson
+---------------------------------------------------------------------------- */
+
+params [
+    ["_unit", objNull, [objNull]],
+    ["_sound", 0, [0]]
+];
+
+if (_unit == objNull) exitWith {false};
```
Comment (commy2) [inline addons/accessibility/fnc_vanillaSoundPlayed.sqf:24]: isNull _unit

File: addons/accessibility/fnc_vanillaSoundPlayed.sqf [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    // Stabilizing
+    case 11: {["stabilizing", []]}; // Not sure what this sounds like, may need to be added
+    // Healing
+    case 12: {["healing", ["EQUIPMENT"]]};
+    // Healing with Medkit
+    case 13: {["healing_with_medkit", ["EQUIPMENT"]]};
+    // Recovered
+    case 14: {["recovered", []]}; // Not sure what this sounds like, may need to be added
+    // Breath Held
+    case 15: {["breath_held", ["BREATH"]]};
```
Comment (commy2) [inline addons/accessibility/fnc_vanillaSoundPlayed.sqf:57]: Comments seem superfluous. They just repeat the name of the key.

File: addons/accessibility/fnc_showHints.sqf [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#include "script_component.hpp"
+
+for "_i" from (count GVAR(hints) - 1) to 0 step -1 do {
+    private _hint = GVAR(hints) select _i;
+    _hint params ["_text", "_source", "_expiresAt"];
+
+    if (diag_tickTime > _expiresAt) then {
```
Comment (commy2) [inline addons/accessibility/fnc_showHints.sqf:7]: Probably not savegame compatible, as diag_tickTime (unlike CBA_missionTime) is not strictly monotonically increasing.

File: addons/accessibility/fnc_shouldPlay.sqf [hunk 8]
```diff
@@ -0,0 +1,48 @@
+#include "script_component.hpp"
+/* ----------------------------------------------------------------------------
+Function: CBA_accessibility_fnc_shouldPlay
+
+Description:
+    Checks if a sound should be played based on the categories it is in.
```
Comment (commy2) [inline addons/accessibility/fnc_shouldPlay.sqf:6]: It only checks the cats. It doesn't handle sounds at all :)

"Check if all of the given sound categories are enabled."

File: addons/accessibility/fnc_shouldPlay.sqf [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        case "SCREAM":      {GVAR(soundsScream)};
+        case "FOOD":        {GVAR(soundsFood)};
+        case "EQUIPMENT":   {GVAR(soundsEquipment)};
+        case "REPETITIVE":  {GVAR(soundsRepetitive)};
+        default {
+            WARNING_1("Unknown sound category: %1", _x);
+            false
+        };
+    } == 2
+}) == -1
```
Comment (commy2) [inline addons/accessibility/fnc_shouldPlay.sqf:48]: Hard to read. Maybe break up.

```sqf
    private _enabled = {
        switch (toUpper _this) do {
            case "BREATH":      {GVAR(soundsBreath)};
            case "BODILY":      {GVAR(soundsBodily)};
            case "PAIN":        {GVAR(soundsPain)};
            case "SCREAM":      {GVAR(soundsScream)};
            case "FOOD":        {GVAR(soundsFood)};
            case "EQUIPMENT":   {GVAR(soundsEquipment)};
            case "REPETITIVE":  {GVAR(soundsRepetitive)};
            default {
                WARNING_1("Unknown sound category: %1", _x);
                false
            };
        } != 2 // return
    }

    // enabled for all 
    _categories findIf {!call _enabled} == -1 // return
```

File: addons/accessibility/fnc_shouldHint.sqf [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        case "BREATH":      {GVAR(soundsBreath)};
+        case "BODILY":      {GVAR(soundsBodily)};
+        case "PAIN":        {GVAR(soundsPain)};
+        case "SCREAM":      {GVAR(soundsScream)};
+        case "FOOD":        {GVAR(soundsFood)};
+        case "EQUIPMENT":   {GVAR(soundsEquipment)};
+        case "REPETITIVE":  {GVAR(soundsRepetitive)};
+        default {
+            WARNING_1("Unknown sound category: %1", _x);
+            false
```
Comment (commy2) [inline addons/accessibility/fnc_shouldHint.sqf:47]: false != 0 errors in SQF :)

File: addons/accessibility/fnc_say3D.sqf [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+params [
+    ["_sound", "", [""]],
+    ["_categories", [], [[]]],
+    ["_source", objNull, [objNull, []]],
+    ["_args", [], [[]]],
+    ["_hint", "", ["", []]]
+];
+
+if (_source isEqualTo objNull) exitWith {false};
```
Comment (commy2) [inline addons/accessibility/fnc_say3D.sqf:49]: isNull _source

File: addons/accessibility/fnc_playSound3D.sqf [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+if (_shouldHint) then {
+    if (_hint isEqualTo "") then {
+        _hint = getText (configFile >> "CfgSounds" >> _sound >> "cba_accessibility_hint");
+    };
+    if (_hint isEqualTo "") then {
+        _hint = _sound;
+    };
+    if (_source isEqualType []) then {
```
Comment (commy2) [inline addons/accessibility/fnc_playSound3D.sqf:67]: _source can't be an array, according to function signature / params.

File: addons/accessibility/fnc_vanillaSoundPlayed.sqf (hunk 5 above)
Comment (veteran29) [reply addons/accessibility/fnc_vanillaSoundPlayed.sqf:24]: null == null returns false.

Comment (commy2) [reply addons/accessibility/fnc_vanillaSoundPlayed.sqf:24]: Exactly. The check currently is a paradox.

File: addons/accessibility/fnc_showHints.sqf (hunk 7 above)
Comment (TACHarsis) [reply addons/accessibility/fnc_showHints.sqf:7]: Also not pause compatible, runs through singleplayer pause screen

**Comment by TACHarsis** [comment]
What happened?

**Comment by BrettMayson** [comment]
I just haven't had time to continue this, feel free to take over the PR if you'd like

**Pull request closed without merging**